sw/device/silicon_creator/lib/ownership/isfb.c

```diff
@@ -0,0 +1,124 @@
+  *checks_performed_count = UINT32_MAX;
+  if (ext->header.name != kManifestExtIdIsfb ||
+      (hardened_bool_t)owner_config->isfb == kHardenedBoolFalse) {
+    return kErrorOwnershipISFBNotPresent;
+  }
+  const owner_isfb_config_t *isfb = owner_config->isfb;
+  if (ext->product_expr_count * 2 > isfb->product_words ||
+      ext->product_expr_count > kProductExprMaxCount) {
+    return kErrorOwnershipISFBProductExpCnt;
+  }
+
+  // The ISFB info page configuration is handled in the owner's
+  // `FlashInfoConfig` settings.
+  flash_ctrl_info_page_t isfb_info_page;
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_type0_params_build(
+      isfb->bank, isfb->page, &isfb_info_page));
+
+  // There are in total 128 bits in the strike mask. Each bit corresponds to a
+  // `uint32_t` word.
+  uint32_t strikes[kExpectedStrikeRegionWordsCount];
+  static_assert(sizeof(ext->strike_mask) * CHAR_BIT == kExpectedStrikeBitCount,
+                "Strike mask size mismatch");
+  static_assert(sizeof(strikes) == kExpectedStrikeRegionBytesCount,
+                "Data size mismatch");
+
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_read(&isfb_info_page, /*offset=*/0,
+                                                ARRAYSIZE(strikes), strikes));
+
+  uint32_t strike_cnt_ok = 0;
+  uint32_t strike_cnt_bad = 0;
+  uint32_t *strike_word = (uint32_t *)strikes;
+  size_t i;
+  for (i = 0; i < ARRAYSIZE(strikes); ++i, ++strike_word) {
+    uint32_t strike_bit = ext->strike_mask[i >> 5] & (1 << (i & 31));
+    if (launder32(strike_bit) && *strike_word != UINT32_MAX) {
+      strike_cnt_bad++;
+    } else if (!launder32(strike_bit) || *strike_word == UINT32_MAX) {
+      strike_cnt_ok++;
+    } else {
+      // This condition is equivalent to a potential fault in either the strike
+      // mask or the ISFB info buffer.
+      strike_cnt_bad++;
+    }
+  }
+  // Check loop completion and count consistency.
+  HARDENED_CHECK_EQ(strike_cnt_ok + strike_cnt_bad + i,
+                    kExpectedStrikeBitCount * 2);
+
+  if (launder32(strike_cnt_bad) > 0) {
+    return kErrorOwnershipISFBStrikeMask;
+  }
+
+  uint32_t product_expr[kProductExprMaxCount];
+  static_assert(sizeof(product_expr) == KProductExprMaxCountBytes,
+                "Product expressions data size mismatch");
+
+  // Read the product expressions from the ISFB info page.
+  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_read(
+      &isfb_info_page, /*offset=*/1024, ARRAYSIZE(product_expr), product_expr));
+
+  uint32_t pexpr_cnt_ok = 0;
+  uint32_t pexpr_cnt_bad = 0;
+
+  // The `product_expr_count` boundary check is performed at the beginning of
+  // the function.
+  for (i = 0; i < ext->product_expr_count; ++i) {
+    if ((product_expr[i] & ext->product_expr[i].mask) ==
+        ext->product_expr[i].value) {
+      pexpr_cnt_ok++;
+    } else {
+      pexpr_cnt_bad++;
+    }
+  }
+  HARDENED_CHECK_EQ(pexpr_cnt_ok + pexpr_cnt_bad, ext->product_expr_count);
+  HARDENED_CHECK_EQ(i, ext->product_expr_count);
```

Comment:
Do we want to do 2 checks here, or do we want to do one check that, similar to the last check, is a calculation and a check:
```
HARDENED_CHECK_EQ(pexpr_cnt_ok + pexpr_cnt_bad + i , 2 * ext->product_expr_count);
```

Reply:
Fixed.